lib/services/awsAdapter.js

```diff
@@ -261,6 +272,93 @@ function AWSAdapter (options) {
 		}
 	};
 
+	this.getKeyPair = function (createKeyName, existingKeyName) {
+
+		var keyPairParams;
+
+		var keyPair = {
+			keyLocation : undefined,
+			name        : undefined
+		};
+
+		if (createKeyName === undefined) {
+			if (existingKeyName === undefined) {
+				//Default name and location of the Service Maker key-pair.
+				existingKeyName = "service-maker";
+				keyPair.keyLocation = DEFAULT_BUCKET + existingKeyName + ".pem";
+			}
+
+			keyPair.name = existingKeyName;
+
+			keyPairParams = {
+				KeyNames : [ existingKeyName ]
+			};
+
+			return ec2.describeKeyPairsAsync(keyPairParams)
+			.then(function () {
+				return keyPair;
+			})
+			.catch(function (error) {
+				if (error.name === "InvalidKeyPair.NotFound") {
+
+					keyPairParams = {
+						KeyNames : existingKeyName
+					};
+
+					return ec2.createKeyPairAsync(keyPairParams)
+					.then(function (key) {
+						var params = {
+							Bucket : "service-maker",
+							Key    : existingKeyName + ".pem",
+							Body   : key.KeyMaterial
+						};
+						return s3.uploadAsync(params);
+					})
+					.then(function (result) {
+						keyPair.name = existingKeyName;
+						keyPair.keyLocation = result.Location;
+						return keyPair;
+					})
+					.catch(function (error) {
+						return Bluebird.reject(error);
+					});
+				}
+				else {
+					return Bluebird.reject(error);
+				}
+			});
+		}
+		else if (createKeyName !== undefined && existingKeyName === undefined) {
+			keyPairParams = {
```

Comment:
This code, and the create code above, could be consolidated into one helper function.